lib/iris/fileformats/ugrid_cf_reader.py

```diff
@@ -0,0 +1,195 @@
+
+    * `self.cfreader` : a CFReader object to interpret the CF data from the
+      file for cube creation, while ignoring the UGRID mesh data.
+
+    * `self.complete_ugrid_cube(cube)` a call to add the relevant UGRID
+      information to a cube created from the cfreader data.
+
+    This allows us to decouple UGRID from CF support with minimal changes to
+    the existing `iris.fileformats.netcdf` code, which is intimately coupled to
+    both the CFReader class and the netCDF4 file interface.
+
+    """
+
+    def __init__(self, filename, *args, **kwargs):
+        self.filename = os.path.expanduser(filename)
+        dataset = netCDF4.Dataset(self.filename, mode="r")
+        self.dataset = dataset
+        meshes = {}
+        for meshname in find_mesh_names(self.dataset):
+            mesh = UGrid()
+            load_grid_from_nc_dataset(dataset, mesh, mesh_name=meshname)
+            meshes[meshname] = mesh
+        self.meshes = meshes
+
+        # Generate list of excluded variable names.
+        exclude_vars = list(meshes.keys())
+
+        temp_xios_fix = kwargs.pop("temp_xios_fix", False)
+        if not temp_xios_fix:
+            # This way *ought* to work, but maybe problems with the test file ?
+            for mesh in meshes.values():
+                mesh_var = dataset.variables[mesh.mesh_name]
+                for attr in mesh_var.ncattrs():
+                    if attr in _UGRID_LINK_PROPERTIES:
+                        exclude_vars.extend(mesh_var.getncattr(attr).split())
+        else:
+            # A crude and XIOS-specific alternative ..
+            exclude_vars += [
+                name
+                for name in dataset.variables.keys()
+                if any(name.startswith(meshname) for meshname in meshes.keys())
+            ]
+
+        # Identify possible mesh dimensions and make a map of them.
+        meshdims_map = {}  # Maps {dimension-name: (mesh, mesh-location)}
+        for mesh in meshes.values():
+            mesh_var = dataset.variables[mesh.mesh_name]
+            if mesh.faces is not None:
+                # Work out name of faces dimension and record it.
+                faces_varname = mesh_var.face_node_connectivity
+                faces_var = dataset.variables[faces_varname]
+                if "face_dimension" in faces_var.ncattrs():
+                    faces_dim_index = faces_var.getncattr("face_dimension")
+                else:
+                    faces_dim_index = 0
+                faces_dim_name = faces_var.dimensions[faces_dim_index]
+                meshdims_map[faces_dim_name] = (mesh, "face")
+            if mesh.edges is not None:
+                # Work out name of edges dimension and record it.
+                edges_varname = mesh_var.edge_node_connectivity
+                edges_var = dataset.variables[edges_varname]
+                if "edge_dimension" in edges_var.ncattrs():
+                    edges_dim_index = faces_var.getncattr("face_dimension")
```

Comment:
Looks like this still needs fixing??

Reply:
Sorry, I held off to avoid interrupting Travis, then forgot to push it..